### DISCUSSION TITLE
Prompt length (244 characters or tokens?)

### DISCUSSION THREAD

**Discussion by RafalCer** [openai/whisper#1824] [Q&A]
Hi,

Whisper's prompt is an awesome feature that allows for accurate handling of tricky words, such as proper nouns denoting company names (e.g. _Whatagraph_). However, it is a bit unclear from [the documentation](https://platform.openai.com/docs/guides/speech-to-text/prompting) what the maximum length of the prompt is. As you can see in the screenshot below, the documentation first states that "_Whisper only considers the first 244 tokens of the prompt._"; however, the next paragraph mentions that "_this technique is limited to only 244 characters_".

<img width="418" alt="image" src="https://github.com/openai/whisper/assets/78570551/7589e79d-ac8d-4d68-a60a-f2dbb2226bef">

Has anyone explored whether the prompt is limited to the first 244 **characters** (letters) or **tokens** (units determined by the tokenizer)?

Thank you!

**Comment by glangford** [comment]
Some evidence points to the limit being *224 tokens* but I don't think this has been discussed in the past. Here is some more conflicting information.

FYI from the Whisper prompting guide
> In addition, the prompt is limited to only 224 tokens. If the prompt is longer than 224 tokens, only the final 224 tokens of the prompt will be considered; all prior tokens will be silently ignored. The tokenizer used is the [multilingual Whisper tokenizer](https://github.com/openai/whisper/blob/main/whisper/tokenizer.py#L361).

https://cookbook.openai.com/examples/whisper_prompting_guide

That said, this space is for the open source repo of whisper and it appears you are using the online API. So I would consider asking this question over at the community developer forum for the online version of whisper.

**Reply by RafalCer** [reply]
Thank you for the response! I'm actually using the open source version, but the aforementioned documentation was the only (official) piece of information regarding prompting I could find.

Also, thank you for pointing me to the prompting guide. It consistently mentions that the limit is 244 _tokens_ rather than _characters_. I also had a closer look at the source code and it seems like the prompt is created and truncated only after the prompt input is encoded, so I'm pretty sure that the limit should apply for **tokens**. 

I experimented a bit with Whisper prompting by transcribing an English file with 2 minutes of speech. At around 01:00, a last name is pronounced which Whisper is unable to transcribe properly on its own, but does perfectly well when it is prompted with the spelling of the last name. In my experiments I tried cluttering the prompt with some sentences from Wikipedia (~700 characters, ~140 tokens), and inserting the problematic last name in the beginning, middle, end, and some random positions of the prompt. Interestingly, the prompt only helped if I placed the last name as one of the last ~5 words in the prompt; in other cases, it would continuously misspell the last name even though it was clearly within the non-truncated range. I also tried cutting down the entire prompt to just 1 sentence of ~80 characters, but the last name would still be spelled incorrectly unless I placed it at the end of my prompt. 

I know that there is quite a bit of randomness in Whisper and especially its prompting feature, but I'm curious if anyone has observed or managed to crack how to use the prompting properly.

**Reply by JoaquinKeller** [reply]
https://platform.openai.com/docs/guides/speech-to-text/prompting in the API doc they say:
- «The model will only consider the final 224 tokens of the prompt and ignore anything earlier»
- «It's important to keep in mind that Whisper only considers the first 244 tokens of the prompt»

So is it the final 224 tokens or the first 244 tokens of the prompt that are taken into account?

**Comment by ClaireCJS** [comment]
224 isn't enough. It should at least be enough to contain song lyrics for songs with lots of words.  Some of us are transcribing lyrics to songs into TIMED lyrics, and whisper is awful at accuracy and needs to be primed with the official lyrics of the song.

Songs have more than 224 words in them. Especially rap, which I don't really listen to, but which should be considered.

These files aren't going to be "chunked up"

**Reply by EtienneAb3d** [reply]
@ClaireCJS 
If you have the official lyrics of a song, you can try to map it using the Alignment of Whisper-TS:
https://github.com/jianfch/stable-ts?tab=readme-ov-file#alignment
;-)

**Reply by ClaireCJS** [reply]
> @ClaireCJS If you have the official lyrics of a song, you can try to map it using the Alignment of Whisper-TS: https://github.com/jianfch/stable-ts?tab=readme-ov-file#alignment ;-)

That's not what I'm trying to do, quite specifically because it's the wrong approach.

Most lyrics in most archives are not word-for-word accounts. Choruses are skipped, parts that repeat are not repeated, etc.

For 25 years, I've been downloading lyrics.  To get a word-for-word accurate lyric set out of a downloaded lyric is a 1% thing, not a 100% thing.  1% of the time are they truly word-for-word accurate.

So if you want a song to truly be transcribed, merely aligning the timing downloaded lyrics will give incomplete results 99% of the time.

The ONLY way to get perfect results with real-world data is to use the lyrics to prime the transcriber.

I reiterate: It is the ONLY way.

Fortunately, a lot of songs don't have more than 224 words of lyrics. The priming works pretty good. Results primed with lyrics are really good compared to ones that aren't.

But I'm concerned about what happens when I get to the rap section of my collection. There are a lot more words, said a lot faster, harder to understand/transcribe, and harder to prime if there is a 224 token limit.

Of course I could chunk the file up, but that creates the issue of trying to associate subsections of the lyrics with chunks, and that issue can only really be solved by... listening to the song. At which point we've gone full circle back to a manual transcription process.

I intend to run this on 20,000+ songs...

Admittedly, getting good lyrics requires manual intervention/approval. But once I do that, the rest of the process will be automated. But yea, I'd like >224 to get accurate results for these specific reasons that are based on experience with real-world data.

**Reply by EtienneAb3d** [reply]
@ClaireCJS 
I see how to build a process:
1. transcribe to get a first timed chunking, even if bad words
2. sync the bad transcription with the not perfect official lyrics to get the vocab using https://github.com/EtienneAb3d/WhisperTimeSync
3. split the audio according to text chunks
4. transcribe chunks using prompts fed with the corresponding vocab
5. merge the result to get the final timed transcription

**Reply by RafalCer** [reply]
@ClaireCJS 

I agree that 224 is quite limiting, especially since 224 tokens is definitely less than 224 words. But note that Whisper does not transcribe _any_ duration of audio in one go. Under the hood it chunks the audio into 30 second chunks and processes those. If you play around a bit with the prompt used at inference of every chunk, you could probably fit all the relevant words for that chunk and prime the transcriber properly. Needless to say, some estimation of which words to use in which chunk is needed, and then it's not a given that the priming will lead to a correct transcription.

**Reply by ClaireCJS** [reply]
> @ClaireCJS I see how to build a process:
> 
> 1. transcribe to get a first timed chunking, even if bad words
> 2. sync the bad transcription with the not perfect official lyrics to get the vocab using https://github.com/EtienneAb3d/WhisperTimeSync
> 3. split the audio according to text chunks
> 4. transcribe chunks using prompts fed with the corresponding vocab
> 5. merge the result to get the final timed transcription

So, more about this:

Step #2.

WhisperTimeSync [absolutely removes lines of spoken text](https://github.com/EtienneAb3d/WhisperTimeSync/discussions/8#discussioncomment-11102834) if it can't match them to the TXT file.

Which is exactly what I'm trying to avoid.

And merging is no easy task.

**Reply by EtienneAb3d** [reply]
The goal of step 2 is not to produce a perfect match for all, but to provide you with the correct vocab as many times as it's possible.

**Reply by EtienneAb3d** [reply]
@ClaireCJS 
For each chunk, if you add in the prompt, for example, the last 150 official words found on step 2, and not only the words for this chunk, you should have a good probability to get all the words you need to process this chunk.

**Comment by ClaireCJS** [comment]
That sure is a lot of steps compared to just allowing the acceptance of more tokens 😉

But thanks for pointing out WhisperTimeSync.  I couldn't remember it's name, couldn't find it yesterday when I was looking for that concept, and wasn't sure if that concept had been implemented by someone, or if I had dreamed that it had been implemented (version control is hard on your own mind! lol).  Now I know it's real, it's name, and will check it out 🎉
